ic-factory/src/top_up.rs

```diff
@@ -0,0 +1,148 @@
+use candid::Principal;
+use cycles_minting_canister::{
+    IcpXdrConversionRateCertifiedResponse, NotifyError, NotifyTopUp, DEFAULT_CYCLES_PER_XDR,
+    MEMO_TOP_UP_CANISTER,
+};
+use ic_base_types::{CanisterId, PrincipalId};
+use ic_canister::virtual_canister_call;
+use ledger_canister::{
+    AccountIdentifier, BlockHeight, SendArgs, Subaccount, Tokens, DEFAULT_TRANSFER_FEE,
+    TOKEN_SUBDIVIDABLE_BY,
+};
+use num_traits::ToPrimitive;
+
+use crate::error::FactoryError;
+
+const CYCLE_MINTING_CANISTER: &str = "rkp4c-7iaaa-aaaaa-aaaca-cai";
+
+/// This function calculates the amount required for minting cycles for a canister.
+pub async fn cycles_to_icp(cycles: u64) -> Result<u64, FactoryError> {
```

Comment:
It would be nice to move the math to a separate function from here and add a few more unit tests for this math. I would suggest using icp to cycles conversion function from IC code to write tests like `given N ICP: covert N ICP to cycles with IC code, then convert resulting value backwards with this code, check that the result equals N` for different (also extreme) values of N.